apps/pipelines/nodes/base.py

```diff
@@ -23,71 +20,65 @@ def add_messages(left: list, right: list):
+                input = state["messages"][-1]
+                output = ... # do something
+                return output # The state will be updated with output
 
        class FunLambdaNode(PipelineNode):
             required_parameter_1: str
 
-            def get_runnable(self, node: Node):
-                def fn(input: Input):
-                    return self.required_parameter_1
-                return RunnableLambda(fn, name=self.__class__.__name__)
+            def _process(self, state: PipelineState) -> PipelineState:
+                ...
+                return # The state will not be updated, since None is returned
+
     """
 
+    config: RunnableConfig | None = None
+
     class Config:
         arbitrary_types_allowed = True
 
     @classmethod
     def build(cls, node: Node) -> Callable[[dict], dict]:
         try:
-            built_node = cls(**node.params)
+            return cls(**node.params)
         except ValidationError as ex:
             raise PipelineNodeBuildError(ex)
 
-        return built_node.get_runnable(node)
+    def process(self, state: PipelineState, config) -> PipelineState:
+        self.config = config
+        cls_name = self.__class__.__name__
+        self.logger.info(f"{cls_name} starting")
 
-    @classmethod
-    def get_callable(cls, node: Node) -> Callable:
-        built_node = cls.build(node)
+        output = self._process(state)
 
-        def fn(state: PipelineState) -> PipelineState:
-            if isinstance(built_node, BasePromptTemplate):
-                input = {"input": state["messages"][-1]}
-            else:
-                input = state["messages"][-1]
-            result = built_node.invoke(input)
-            if isinstance(result, BaseMessage):
-                output = result.content
-            elif isinstance(result, StringPromptValue):
-                output = result.text
-            else:
-                output = result
-            return PipelineState(messages=[output])
-
-        return fn
-
-    def get_runnable(self, node: Node) -> Runnable:
-        """Get a predefined runnable to be used in the pipeline"""
+        self.logger.info(f"{cls_name} finished with output: {output}")
+
+        # Append the output to the state, otherwise do not change the state
+        return PipelineState(messages=[output]) if output else PipelineState()
+
+    def _process(self, state: PipelineState) -> PipelineState:
+        """The method that executes node specific functionality"""
         raise NotImplementedError
 
-    def logger(self, config):
-        for handler in config["callbacks"].handlers:
+    @cached_property
+    def logger(self):
+        for handler in self.config["callbacks"].handlers:
             if isinstance(handler, PipelineLoggingCallbackHandler):
                 return handler.logger
-        raise AttributeError("No logger found")
+
+    def experiment_session(self, state: PipelineState) -> ExperimentSession:
```

Comment:
nit: this could go on the state object itself

Reply:
https://github.com/dimagi/open-chat-studio/pull/474/commits/e5b09b58d8265b8aa51c146ff260cc27be54525a